flocker/ca/functional/test_validation.py

```diff
@@ -0,0 +1,274 @@
+
+"""
+Test validation of keys generated by flocker-ca.
+"""
+
+from tempfile import mkdtemp
+from shutil import rmtree
+import atexit
+
+from pyrsistent import PRecord, field
+
+from OpenSSL.SSL import Error as SSLError
+
+from twisted.trial.unittest import TestCase
+from twisted.internet.ssl import PrivateCertificate
+from twisted.internet.endpoints import (
+    SSL4ServerEndpoint, connectProtocol, SSL4ClientEndpoint,
+    )
+from twisted.internet import reactor
+from twisted.internet.defer import Deferred
+from twisted.python.filepath import FilePath
+from twisted.internet.protocol import Protocol, ServerFactory
+
+
+from ...testtools import find_free_port
+from .._ca import (
+    RootCredential, UserCredential, NodeCredential, ControlCredential)
+from .._validation import (
+    ControlServicePolicy, amp_server_context_factory, rest_api_context_factory,
+    )
+
+
+EXPECTED_STRING = b"Mr. Watson, come here; I want to see you."
+
+
+class SendingProtocol(Protocol):
+    """
+    Send a string.
+    """
+    def connectionMade(self):
+        self.transport.write(EXPECTED_STRING)
+        self.transport.loseConnection()
+
+
+class ReceivingProtocol(Protocol):
+    """
+    Expect a string.
+
+    :ivar Deferred result: Fires on receiving response or if disconnected
+         before that.
+    """
+    def __init__(self):
+        self.result = Deferred()
+        self._buffer = b""
+
+    def dataReceived(self, data):
+        self._buffer += data
+        if self._buffer == EXPECTED_STRING:
+            self.result.callback(None)
+            self.result = None
+
+    def connectionLost(self, reason):
+        if self.result:
+            self.result.errback(reason)
+            self.result = None
+
+
+class CredentialSet(PRecord):
+    """
+    A full set of credentials for a CA.
+
+    :param RootCredential root: The CA root credential.
+    :param ControlCredential control: A control service credential.
+    :param UserCredential user: A user credential.
+    :param NodeCredential node: A CA root credentials.
+    """
+    root = field()
+    control = field()
+    user = field()
+    node = field()
+
+    @staticmethod
+    def create():
+        """
+        :return: A new ``CredentialSet``.
+        """
+        directory = FilePath(mkdtemp())
+        atexit.register(rmtree, directory.path)
+        root = RootCredential.initialize(directory, b"mycluster")
+        user = UserCredential.initialize(directory, root, u"allison")
+        node = NodeCredential.initialize(directory, root)
+        control = ControlCredential.initialize(directory, root, b"127.0.0.1")
+        return CredentialSet(root=root, user=user, node=node, control=control)
+
+
+def get_sets():
+    """
+    :return: Tuple, a pair of ``CredentialSet`` instances.
+    """
+    a = CredentialSet.create()
+    b = CredentialSet.create()
+    global get_sets
+
+    def get_sets():
+        return a, b
+    return get_sets()
```

Comment:
`a` and `b` (with better names :smile: ) could be the globals.  I think that might be less confusing to maintainers than a function that replaces itself.